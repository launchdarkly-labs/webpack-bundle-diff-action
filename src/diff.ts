// Mostly based off of https://github.com/ZachGawlik/webpack-stats-diff/tree/master/src with better
// support for contenthash, typescript, etcâ€¦

import type { StatsCompilation, StatsAsset } from 'webpack';
import type { BundleAnalyzerPlugin } from 'webpack-bundle-analyzer';

// We tweak the Webpack stats type to make,
//   - assetsByChunkName and assets required
//   - only require name and size on assets
type Modify<T, R> = Omit<T, keyof R> & R;

type WebpackAsset = Pick<StatsAsset, 'name' | 'size'>;

type WebpackStats = Modify<
  StatsCompilation,
  {
    assets: WebpackAsset[];
  }
>;

type FunkyAsset = {
  name: string;
  gzipSize: number;
  statSize: number;
  parsedSize: number;
};

export type AssetDiff = {
  name: string;
  baseSize: number;
  headSize: number;
  delta: number;
  ratio: number;
};

export type Diff = {
  added: AssetDiff[];
  removed: AssetDiff[];
  bigger: AssetDiff[];
  smaller: AssetDiff[];
  unchanged: AssetDiff[];
};

const DEFAULT_DIFF_THRESHOLD = 0.05; // 5%

const ASSET_NAME_REGEXP = /^(?<assetname>[a-zA-Z0-9\.\-_]+)\.([a-zA-Z0-9]{20})\.(?<extension>js|css)$/;

const parseAssetName = (name: string) => {
  const match = name.match(ASSET_NAME_REGEXP);

  if (!match || !match.groups) {
    return;
  }

  return {
    ...match.groups,
    canonicalName: `${match.groups.assetname}.${match.groups.extension}`,
  };
};

export function getDiff(
  analysis: {
    base: { stats: WebpackStats; report: BundleAnalyzerPlugin.JsonReport };
    head: { stats: WebpackStats; report: BundleAnalyzerPlugin.JsonReport };
  },
  { diffThreshold = DEFAULT_DIFF_THRESHOLD }: { diffThreshold?: number } = {},
): Diff {
  const byName: {
    base: Record<string, FunkyAsset>;
    head: Record<string, FunkyAsset>;
  } = {
    base: Object.fromEntries(
      analysis.base.report
        .map((item) => {
          const parsed = parseAssetName(item.label);

          if (!item.isAsset || !parsed) {
            return [];
          }

          return [
            parsed.canonicalName,
            {
              name: parsed.canonicalName,
              gzipSize: item.gzipSize,
              statSize: item.statSize,
              parsedSize: item.parsedSize,
            },
          ];
        })
        // filter out assets that didn't match for some reason
        .filter((entry) => entry.length !== 0),
    ),
    head: Object.fromEntries(
      analysis.head.report
        .map((item) => {
          const parsed = parseAssetName(item.label);

          if (!item.isAsset || !parsed) {
            return [];
          }

          return [
            parsed.canonicalName,
            {
              name: parsed.canonicalName,
              gzipSize: item.gzipSize,
              statSize: item.statSize,
              parsedSize: item.parsedSize,
            },
          ];
        })
        // filter out assets that didn't match for some reason
        .filter((entry) => entry.length !== 0),
    ),
  };

  let diff: Record<
    'added' | 'removed' | 'bigger' | 'smaller' | 'unchanged',
    AssetDiff[]
  > = {
    added: [],
    removed: [],
    bigger: [],
    smaller: [],
    unchanged: [],
  };

  for (let name of Object.keys(byName.base)) {
    const baseAsset = byName.base[name];
    const baseSize = baseAsset.parsedSize;

    const headAsset = byName.head[name];

    // Removed
    if (!headAsset) {
      const headSize = 0;
      diff.removed.push({
        name,
        baseSize: baseAsset.parsedSize,
        headSize,
        delta: -baseSize,
        ratio: -1,
      });
    } else {
      const headSize = headAsset.parsedSize;
      const delta = headSize - baseSize;
      const ratio = (1 - headSize / baseSize) * -1 || 0;
      const d: AssetDiff = {
        name,
        baseSize,
        headSize,
        ratio,
        delta,
      };

      if (ratio > diffThreshold) {
        diff.bigger.push(d);
      } else if (ratio < -1 * diffThreshold) {
        diff.smaller.push(d);
      } else {
        diff.unchanged.push(d);
      }
    }
  }

  for (let name of Object.keys(byName.head)) {
    const headAsset = byName.head[name];
    const headSize = headAsset.parsedSize;

    const baseAsset = byName.base[name];

    // Added
    if (!baseAsset) {
      diff.added.push({
        name,
        baseSize: 0,
        headSize,
        delta: headSize,
        ratio: 1,
      });
    }
  }

  return diff;
}
