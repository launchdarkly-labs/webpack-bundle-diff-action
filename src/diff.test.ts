import { getDiff } from './diff';
import {
  renderSection,
  renderAddedTable,
  renderBiggerTable,
  renderRemovedTable,
  renderSmallerTable,
  renderUnchangedTable,
  renderSummaryTable,
  renderReductionCelebration,
  pluralize,
} from './render';

const diff = getDiff({
  base: {
    stats: require('../base-webpack-stats.json'),
    report: require('../base-webpack-bundle-analyzer-report.json'),
  },
  head: {
    stats: require('../head-webpack-stats.json'),
    report: require('../head-webpack-bundle-analyzer-report.json'),
  },
});

test('section', () => {
  expect(
    renderSection({
      title: `⚠️ ${diff.bigger.length} ${pluralize(
        diff.bigger.length,
        'bundle',
        'bundles',
      )} got bigger`,
      isEmpty: diff.bigger.length === 0,
      children: renderBiggerTable({ assets: diff.bigger }),
    }),
  ).toMatchSnapshot();
});

test('empty section', () => {
  expect(
    renderSection({
      title: `⚠️ ${diff.bigger.length} ${pluralize(
        diff.bigger.length,
        'bundle',
        'bundles',
      )} got bigger`,
      isEmpty: true,
      children: renderBiggerTable({ assets: [] }),
    }),
  ).toMatchSnapshot();
});

test('added diff', () => {
  expect(renderAddedTable({ assets: diff.added })).toMatchSnapshot();
});

test('removed diff', () => {
  expect(renderRemovedTable({ assets: diff.removed })).toMatchSnapshot();
});

test('bigger diff', () => {
  expect(renderBiggerTable({ assets: diff.bigger })).toMatchSnapshot();
});

test('smaller diff', () => {
  expect(renderSmallerTable({ assets: diff.smaller })).toMatchSnapshot();
});

test('unchanged diff', () => {
  expect(renderUnchangedTable({ assets: diff.unchanged })).toMatchSnapshot();
});

test('summary', () => {
  expect(renderSummaryTable({ diff })).toMatchSnapshot();
});

test('reduction celebration', () => {
  expect(renderReductionCelebration({ diff })).toMatchSnapshot();
});
