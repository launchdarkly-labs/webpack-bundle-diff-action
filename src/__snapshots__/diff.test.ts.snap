// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`added diff 1`] = `
"| Asset                             |   Size ▾ |
| :-------------------------------- | -------: |
| \`ApprovalRequestContainer.js\`     | \`719 kB\` |
| \`ManageFlagTargetingContainer.js\` | \`647 kB\` |"
`;

exports[`bigger diff 1`] = `
"| Asset                              | Base size | Head size |  Delta ▾ |  Delta % |
| :--------------------------------- | --------: | --------: | -------: | -------: |
| \`ManageFlagVariationsContainer.js\` |   \`35 kB\` |  \`682 kB\` | \`647 kB\` | \`1,861%\` |
| \`common.js\`                        |  \`268 kB\` |  \`544 kB\` | \`276 kB\` |   \`103%\` |
| \`FlagExperimentsContainer.js\`      |  \`203 kB\` |  \`240 kB\` |  \`37 kB\` |    \`18%\` |"
`;

exports[`empty section 1`] = `
"
#### ⚠️ 3 bundles got bigger
No relevant changes.
"
`;

exports[`reduction celebration 1`] = `undefined`;

exports[`removed diff 1`] = `
"| Asset | Size ▾ |
| :---- | -----: |"
`;

exports[`section 1`] = `
"
#### ⚠️ 3 bundles got bigger
| Asset                              | Base size | Head size |  Delta ▾ |  Delta % |
| :--------------------------------- | --------: | --------: | -------: | -------: |
| \`ManageFlagVariationsContainer.js\` |   \`35 kB\` |  \`682 kB\` | \`647 kB\` | \`1,861%\` |
| \`common.js\`                        |  \`268 kB\` |  \`544 kB\` | \`276 kB\` |   \`103%\` |
| \`FlagExperimentsContainer.js\`      |  \`203 kB\` |  \`240 kB\` |  \`37 kB\` |    \`18%\` |
"
`;

exports[`smaller diff 1`] = `
"| Asset                                |  Base size | Head size |   Delta ▾ | Delta % |
| :----------------------------------- | ---------: | --------: | --------: | ------: |
| \`ManageFlagContainer.js\`             | \`1,016 kB\` |   \`30 kB\` | \`-985 kB\` |  \`-97%\` |
| \`users.entry.js\`                     | \`1,012 kB\` |  \`737 kB\` | \`-276 kB\` |  \`-27%\` |
| \`segments.entry.js\`                  | \`1,170 kB\` |  \`896 kB\` | \`-274 kB\` |  \`-23%\` |
| \`integrations.entry.js\`              |   \`569 kB\` |  \`364 kB\` | \`-205 kB\` |  \`-36%\` |
| \`AccountSettingsContainer.js\`        |   \`350 kB\` |  \`150 kB\` | \`-200 kB\` |  \`-57%\` |
| \`FlagComparisonContainer.js\`         |   \`471 kB\` |  \`290 kB\` | \`-181 kB\` |  \`-39%\` |
| \`FlagToggleConfirmationContainer.js\` |   \`813 kB\` |  \`648 kB\` | \`-165 kB\` |  \`-20%\` |
| \`FlagCreatorContainer.js\`            |   \`871 kB\` |  \`713 kB\` | \`-158 kB\` |  \`-18%\` |
| \`DashboardContainer.js\`              |   \`273 kB\` |  \`162 kB\` | \`-111 kB\` |  \`-41%\` |
| \`metrics.entry.js\`                   |   \`183 kB\` |   \`97 kB\` |  \`-87 kB\` |  \`-47%\` |
| \`eventDebugger.entry.js\`             |   \`178 kB\` |   \`96 kB\` |  \`-82 kB\` |  \`-46%\` |
| \`ManageProfileContainer.js\`          |   \`267 kB\` |  \`235 kB\` |  \`-32 kB\` |  \`-12%\` |
| \`auditLog.entry.js\`                  |   \`196 kB\` |  \`164 kB\` |  \`-32 kB\` |  \`-16%\` |
| \`metricsDetails.js\`                  |    \`60 kB\` |   \`32 kB\` |  \`-28 kB\` |  \`-47%\` |
| \`ArchiveFlagPanelContainer.js\`       |    \`54 kB\` |   \`35 kB\` |  \`-19 kB\` |  \`-36%\` |
| \`oauth2Authorize.entry.js\`           |    \`85 kB\` |   \`67 kB\` |  \`-18 kB\` |  \`-21%\` |
| \`ManageRolesContainer.js\`            |    \`56 kB\` |   \`40 kB\` |  \`-16 kB\` |  \`-29%\` |
| \`ConnectFlagContainer.js\`            |    \`26 kB\` |   \`10 kB\` |  \`-16 kB\` |  \`-60%\` |
| \`ManageProjects.js\`                  |   \`185 kB\` |  \`172 kB\` |  \`-12 kB\` |   \`-7%\` |
| \`NotificationCenterContainer.js\`     |    \`86 kB\` |   \`75 kB\` |  \`-12 kB\` |  \`-14%\` |
| \`FlagAcrossEnvironmentsContainer.js\` |    \`10 kB\` |    \`8 kB\` |   \`-2 kB\` |  \`-16%\` |"
`;

exports[`summary 1`] = `
"|           |       Delta |
| :-------- | ----------: |
| Bigger    |   \`+960 kB\` |
| Smaller   | \`-2,911 kB\` |
| Added     | \`+1,366 kB\` |
| Removed   |     \`+0 kB\` |
| **Total** |   \`-586 kB\` |"
`;

exports[`unchanged diff 1`] = `
"| Asset                             |   Size ▾ |
| :-------------------------------- | -------: |
| app.js                            | 1,203 kB |
| unauthenticated.js                |   515 kB |
| framework.js                      |   413 kB |
| TimeSeriesComponent.js            |   379 kB |
| CreateOrJoinContainer.js          |   346 kB |
| quickstart.entry.js               |   343 kB |
| ManageTeamContainer.js            |   314 kB |
| AcceptInvitation.js               |   316 kB |
| newHires.js                       |   300 kB |
| codemirror.js                     |   226 kB |
| PasswordContainer.js              |   213 kB |
| CanduPortal.js                    |   209 kB |
| VerifyMFA.js                      |   205 kB |
| FlagInsightsContainer.js          |   175 kB |
| UsageContainer.js                 |   172 kB |
| EnvironmentSwitcher.js            |   163 kB |
| FlagHistoryContainer.js           |   160 kB |
| ManageAuthorization.js            |   149 kB |
| ManageBilling.js                  |   143 kB |
| FlagCodeReferenceListContainer.js |   112 kB |
| RulesOfBehavior.js                |   107 kB |
| pubnub.js                         |    91 kB |
| ManageFlagSettingsContainer.js    |    92 kB |
| ForgotPasswordContainer.js        |    83 kB |
| GlobalHotKeysContainer.js         |    72 kB |
| HotKeys.js                        |    65 kB |
| ManageSecurity.js                 |    64 kB |
| DateRangePickerComponent.js       |    57 kB |
| DatePickerComponent.js            |    55 kB |
| CustomerLogoGrid.js               |    46 kB |
| RuleExclusionContainer.js         |    39 kB |
| newHires.entry.js                 |    36 kB |
| PendingChangesDrawerContainer.js  |    35 kB |
| LoginContainer.js                 |    32 kB |
| OAuthProviderConfig.entry.js      |    26 kB |
| ThirdPartyWidgets.js              |    13 kB |
| runtime.js                        |    10 kB |
| AppNavContainer.js                |     8 kB |
| flags.entry.js                    |     4 kB |
| errorPage.js                      |     0 kB |"
`;
